### ISSUE TITLE
Support "Auto Flush"

### ISSUE THREAD

**Issue by parisholley** [mikro-orm/mikro-orm#2359]
Not to be confused with what appears to be the `autoFlush` option in past version, but how hibernate implements it here:

https://docs.jboss.org/hibernate/orm/5.1/userguide/html_single/chapters/flushing/Flushing.html

In particular, this feature: `prior to executing a JPQL/HQL query that overlaps with the queued entity actions`

There are parts of my service layer which can touch many models at once and multiple inserts/updates, and depending on the order of things, it isn't always appropriate (for performance purposes) to introduce a manual flush, because you aren't sure how upstream services will consume the method running a query. This is particularly true for "long" transactions with a lot of business logic.

Example:

Service A
```typescript

async function parent(){
  const newModel = new Model();
  
  em.persist(newModel);

  await this.serviceb.child();
}
```

Service B
```typescript
async function child(){
  const activity = await em.find(Model, {})
}
```

I would expect that if any models are persisted, any non-identity query (whether using find or createQueryBuilder), would automatically flush the query'd model prior to executing.

**Comment by parisholley** [comment]
here is a crude monkey patch example (in one of my test cases) of how it could work:

```typescript
const persist = em.persist;

em.persist = (obj) => {
  actions.add(obj.constructor.name);

  return persist.call(em, obj);
};

const find = em.find;

em.find = async (...args) => {
  if (actions.has(args[0].constructor.name)) {
    await em.flush();
  }

  return find.apply(em, args);
};
```

a better/smarter of that implementation, would only flush the relevant model (and any cascades) instead of the entire session

**Comment by B4nan** [comment]
Interesting! I guess this should work with all operations, not just insertions, right? 

Sounds like a reasonable default, if it could make it to v5.0. The implementation could be also quite easy.

**Comment by parisholley** [comment]
i updated my example above, one potential gotcha in how i do PK detection is it wont work for composite keys. given that the where clause on `findOne` and doing a lookup by a "key object" could not be distinguished. i don't use composite keys so won't put much thought into that, but something to consider

**Comment by B4nan** [comment]
Do you know if hibernate is actually doing that "partial flush"? The docs look like they flush the whole session all the time.

**Comment by parisholley** [comment]
not sure, it has been a long time since i've worked in Java :)

probably doesn't matter with v1 of the feature, because you'd have to weigh the benefit of (digging through the session models and determine if there are any cascaded entities, or some bi-directional logic) which cause CPU time, vs. just committing to DB.

**Comment by B4nan** [comment]
Yeah, that part would definitely have a performance penalty and added complexity of the implementation, let's keep it simple for now.

**Comment by parisholley** [comment]
i check extends because I use discriminators and often query by the abstract class and use switch cases based on the implementation.

**Comment by parisholley** [comment]
i tried to think about the joined relation bit, only scenario i could see where maybe it gets tripped up, when adding an unpersisted entry into a one-to-many. so long as there isn't a case where you could run a future query, make changes to the query results, then later flush and the original collection overwrites or does something funky. i don't think such a risk would appear for many-to-one.

**Comment by B4nan** [comment]
It feels like a safe way to implement this would be to compute all the change sets, as `persist` needs to be called only for new entities (it is called on managed entities during flush too), and is not called for removed ones at all. We could cache the results if we don't actually flush so we don't compute it twice.

Regarding the relations, I can imagine problems with cascading:

```ts
const book = await em.find(Book, 1);
book.author = new Author();
const authors = await em.find(Author, {}); // if we ignore author relation, we would not auto flush here
```

**Comment by parisholley** [comment]
**first example:**

i could see that, though i'm in the camp of "always persist", vs relying on implicit cascades.. to some extent, i'd prefer the flush mechanism to throw an exception if it hit an object that hadn't explicitly been persisted. for my project, it is rare to generate multiple associated models at once, so i'd prefer to opt-out of the CPU penalty of "cascade scanning" (if there is a penalty)

**second example:**

wouldn't `await em.find(Book, { author: { name: 'abc' } });` realize that the author is already in the identity map and swap it out? if not, this scares me :) but perhaps that is a new issue to discuss (eg: if versioning is enabled on an entity, if they. match, replace, otherwise throw a conflict exception). i really don't ever want someone to query from db and lose unflushed updates.

**Comment by B4nan** [comment]
> (if there is a penalty)

The flush would still happen later, and if we cache the change sets, it is just about _when_, not really a penalty.

> wouldn't await em.find(Book, { author: { name: 'abc' } }); realize that the author is already in the identity map and swap it out? if not, this scares me :)

This is not about swapping entity instance, you would not find the book, as it's author has the old name, until we auto flush (but book entity was not changed). 

> i really don't ever want someone to query from db and lose unflushed updates.

This should never happen, in v4 you explicitly need to pass `FindOptions.refresh: true` to reload such entities including their state (so you can only loose the state if you ask for it explicitly). 

v5 is smarter thanks to this #2263 - there is a heuristic whether we should reload entities, and loaded data are always merged to the entity, with exception of properties that were changed since the entity was loaded - so you get the best of both approaches, no need for explicit refreshing, but never loose something you already changed.

**Comment by parisholley** [comment]
oh sorry, i misread the query (eg: abc search)..i could see this being a more common case with boolean flags and the like. it would seem the only way to solve it is either a) eat the flush cost by include the referenced entities, or b) eat the potential cpu cost of tracking "paths", eg: ['author','name'] and only refresh if that path is hit

w/r/t tp refresh: 👍

**Comment by parisholley** [comment]
per our conversion in nestjs, a complication can be parallel flushes...

const model = new MyModel();
em.persist(model);

await Promise.all([
  this.servicea.work(em),
  this.serviceb.work(em),
]);

if service A and B both do a query on MyModel, could introduce race conditions. one solution could be some type of event-loop tick magic (kind of how the DataLoader project works), where the queries aren't evaluated until the next event loop cycle, that way we can dedupe or batch somehow

**Comment by B4nan** [comment]
I don't see that as a complication, you need to use separate forks to be able to flush in parallel. While autoflushing might trigger flush on find/findOne, you would still flush later even without the autoflush, so you still need separate forks to allow it.

**Comment by parisholley** [comment]
i'm not sure i follow. if have an N+1 behavior downstream, eg:

```typescript
em.transaction((tx) => {
  const entities:MyModel[]; = await em.find(MyModel, {});
  entities[0].foo = 'bar';
  
  await Promise.all(entities.map(async (entity) => {
    const secondaryQuery = await em.find(MyModel, {baz: 'bum'});
  });
});
```

each `find` within the Promises could both see a "dirty" state and result in an autoflush. the `secondaryQuery` logic could be encapsulated in a service layer, potentially multiple layers deep, so you really won't be able to know if you need to be forking things or not.

**Comment by B4nan** [comment]
I see, still this was always a non goal for me personally, never really suggested using `Promise.all` anywhere, I believe that in most cases you can get better results without it (true performance comes from firing less queries, not making them run parallel). Not really a friend with GQL and dataloaders either - if you wanna see that happen, you will need to contribute :]

**Comment by parisholley** [comment]
will also need to watch for `Collection.loadItems()` since it may query for entities which are dirty

**Comment by parisholley** [comment]
something i ran into with my auto-flush feature:

```typescript
const newEntity = em.assign(MyEntity, {
  manyToOne: 1
});

// assume this method runs a query which would force a flush
// if MyEntity.foobar is a required field, i will get a non-null exception
const value = service.calculateSomething(em);

// this would never be reached
newEntity.foobar = value;

em.persist(newEntity);
```

whether using `em.assign` or `em.create`, if your entity gets associated with something in the identity map and you weren't done building the object, errors can occur. of course, simple fix is to just move the service call up, but it makes me wonder if the propagation should not happen until you explicitly do a `em.persist` (or add via a collection)

**Comment by B4nan** [comment]
Well, if you ask me, you should not have entity in invalid state. The required props should be constructor parameters and it should be used directly instead of em.create() (it would use the ctor anyway, so you might as well validate this from inside ctor and keep using it). Its a matter of preference, but this style will mitigate such issues. And em.assign wont get you to such problems, only if you would try to unset required field explicitly.

**Comment by B4nan** [comment]
I made some progress on this. I didn't like the idea of requiring `em.persist()` for managed entities, as well as I saw how slow flushing can be when not needed, so I ended up using a bit smarter approach. All entity properties are defined as get/set, so we know which entity is touched by the user. Then the `tryFlush` check just iterates thru the identity map and checks this boolean flag on each entity. This way we only trigger the flush if some entity actually changed.

Need to spend more time testing this and polish the code, but already got a WIP and most tests are passing.
